File: frontend/src/util/views/riddle.view.ts
export interface Riddle {
  id: number
  title: string
  imageUrl: string
  solved: boolean
  description?: string
  hint?: string
  firstSolver?: string
  creator?: string
  skipPermitted: boolean
}

export interface RiddleWithSolution extends Riddle {
  solution: string
}

export interface RiddleCategoryHistory {
  categoryName: string
  submissions: RiddleWithSolution[]
}

export interface RiddleCategory {
  categoryId: number
  title: string
  completed: number
  total: number
  nextRiddle?: number
}

export enum RiddleSubmissionStatus {
  CORRECT = 'CORRECT',
  WRONG = 'WRONG',
  CANNOT_SKIP = 'CANNOT_SKIP',
  SUBMITTER_BANNED = 'SUBMITTER_BANNED'
}

export interface RiddleSubmissionResult {
  status: RiddleSubmissionStatus
  nextId?: number
}

export interface Hint {
  hint: string
}

File: backend/src/main/kotlin/hu/bme/sch/cmsch/component/riddle/RiddleView.kt
package hu.bme.sch.cmsch.component.riddle

import com.fasterxml.jackson.annotation.JsonView
import hu.bme.sch.cmsch.dto.FullDetails

data class RiddleView(

    @field:JsonView(FullDetails::class)
    var imageUrl: String = "",

    @field:JsonView(FullDetails::class)
    var title: String = "",

    @field:JsonView(FullDetails::class)
    var hint: String? = null,

    @field:JsonView(FullDetails::class)
    var solved: Boolean = false,

    @field:JsonView(FullDetails::class)
    var skipped: Boolean = false,

    @field:JsonView(FullDetails::class)
    var creator: String? = null,

    @field:JsonView(FullDetails::class)
    var firstSolver: String? = null,

    @field:JsonView(FullDetails::class)
    var description: String = "",

    @field:JsonView(FullDetails::class)
    var skipPermitted: Boolean,
)

data class RiddleViewWithSolution(

    @field:JsonView(FullDetails::class)
    var imageUrl: String = "",

    @field:JsonView(FullDetails::class)
    var title: String = "",

    @field:JsonView(FullDetails::class)
    var hint: String? = null,

    @field:JsonView(FullDetails::class)
    var solved: Boolean = false,

    @field:JsonView(FullDetails::class)
    var skipped: Boolean = false,

    @field:JsonView(FullDetails::class)
    var creator: String? = null,

    @field:JsonView(FullDetails::class)
    var firstSolver: String? = null,

    @field:JsonView(FullDetails::class)
    var solution: String = "",

    @field:JsonView(FullDetails::class)
    var description: String = "",
)
